Smt/Reconstruct/Int/Polynorm.lean

```diff
@@ -111,20 +166,93 @@ def mul (p q : Polynomial) : Polynomial :=
 def denote (ctx : Context) (p : Polynomial) : Int :=
   p.foldl (fun acc m => acc + m.denote ctx) 0
 
-theorem denote_neg {p : Polynomial} : p.neg.denote ctx = -p.denote ctx :=
-  sorry
+theorem foldl_add_insert (ctx : Context) :
+  List.foldl (fun z a => z + (Monomial.denote ctx a)) 0 (add.insert m p) =
+  (Monomial.denote ctx m) + List.foldl (fun z a => z + (Monomial.denote ctx a)) 0 p := by
+  induction p with
+  | nil => simp [add.insert]
+  | cons n p ih =>
+    simp only [add.insert]
+    split
+    <;> simp only [List.foldl_cons, Int.add_comm 0, Monomial.foldl_assoc Int.add_assoc]
+    case cons.inr h =>
+      split
+      case inl h1 =>
+        split
+        case inl h2 =>
+          rw [←Int.add_assoc, Int.add_comm, ←Monomial.denote_add h1]
+          simp [Monomial.denote, h2]
+        case inr h2 =>
+          simp [-Int.add_zero, Int.add_comm 0, Monomial.foldl_assoc Int.add_assoc, Monomial.denote_add h1, Int.add_assoc]
+      case inr h1 =>
+        simp only [List.foldl_cons, Int.add_comm 0, ih, Monomial.foldl_assoc Int.add_assoc]
+        rw [←Int.add_assoc, Int.add_comm (Monomial.denote ctx n), Int.add_assoc]
```

Comment:
Proof structure is fine. Numeric names (`h1` and `h2`) are generally not acceptable. `split` can be combined with `rename_i` to avoid using the `case` tactic (you can keep using `case` tactic if you think it's more appropriate here).
```suggestion
  induction p with
  | nil => simp [add.insert]
  | cons n p ih =>
    simp only [add.insert]
    split <;> rename_i hlt <;> simp only [List.foldl_cons, Int.add_comm 0, Monomial.foldl_assoc Int.add_assoc]
    · split <;> rename_i heq
      · split <;> rename_i hneq
        · rw [←Int.add_assoc, Int.add_comm, ←Monomial.denote_add heq]
          simp [Monomial.denote, hneq]
        · simp [-Int.add_zero, Int.add_comm 0, Monomial.foldl_assoc Int.add_assoc, Monomial.denote_add, heq, Int.add_assoc]
      · simp only [List.foldl_cons, Int.add_comm 0, ih, Monomial.foldl_assoc Int.add_assoc]
        rw [←Int.add_assoc, Int.add_comm (Monomial.denote ctx n), Int.add_assoc]
```